### PR TITLE
Transactional build

### LINKED ISSUE #18: Builds are not transactional
We should build in temp dir and `mv` to the esy's store when build is finished.

### PR DESCRIPTION
**Pull request by andreypopp**
Depends on #25. Fixes #18.

Build into `$prefix/_insttmp` first then mv it into `$prefix/_install`. Not that we `_insttmp` and `_install` has the same length, this is important cause we do path replacement in all files before doing `mv`. This is to fight not relocatable compiler toolchain.

### CODE REVIEW

File: lib/esyBuildRuntime.sh [hunk 1]
```diff
@@ -70,20 +70,22 @@ esy-copy-permissions () {
...
-  SRC_STRING="$3"
-  DEST_STRING="$4"
-  # TODO: This uses bbe command which is not available by default on linux,
-  # darwin at least.
-  bbe -e "s|$SRC_STRING|$DEST_STRING|" -o "$OUTPUT_FILE" "$INPUT_FILE"
+  FILE="$1"
+  SRC_STRING="$2"
+  DEST_STRING="$3"
+  # TODO: get rid of python here
+  python -c "
```
Comment (andreypopp) [inline lib/esyBuildRuntime.sh:77]: This introduces dependency on `python` which is not so bad in terms of availability (most systems has it out of the box) but bad in terms of speed.

**Comment by trishume** [comment]
Note that unlike Nix's method, with this it is possible to manufacture collisions by intentionally including the correct `_insttmp` path in a package. Whereas Nix uses an unpredictable cryptographic hash.

I'm not sure if this will be an issue in practice or if there are any possible exploits, but something to think about.

**Comment by andreypopp** [comment]
@trishume good point, we also have few other places where collision is possible, like `cur` prefix in env vars.

**Comment by jordwalke** [comment]
Looks good as well. Nice solution! We should consider marking packages as relocatable to avoid the string replace. I know it's not feasible for existing packages, but for ones we build going forward, I'm sure we will strive for relocatability. Merge when ready.

**Pull request merged**